enhancements/testing/improved-platform-tests.md

```diff
@@ -0,0 +1,456 @@
+        end
+    end
+    loop over all tests
+        alt embedded test?
+            O->>+O: run #lt;test name#gt;#10;
+        else external test?
+            rect rgb(177, 177, 177)
+                O->>+E: run #lt;test name#gt;
+                E->>+O: results
+            end
+        end
+    end
+    O->>+U: results
+-->
+
+### API Extensions
+
+N/A
+
+### Implementation Details/Notes/Constraints
+
+To achieve the first step of the proposal, we need to:
+1. Provide a library for building `<binary>-tests` which exposes two commands:
+   - `list` - responsible for listing tests in JSON format;
+   - `run` - run a single test, returning results in ginkgo compatible format;
+2. Provide a mechanism for labeling tests, similarly how we currently indicate
+   that a test is part of a special group. See [origin/rules.go](https://github.com/openshift/origin/blob/cc42164781a728b804d0bb07f09cf878ec8f7807/test/extended/util/annotate/rules.go)
+   or [kubernetes/rules.go](https://github.com/openshift/kubernetes/blob/8bd5514be14de7fb3105962c8fc206ef5e921a0b/openshift-hack/e2e/annotate/rules.go),
+   in combination with [the annotation mechanism](https://github.com/openshift/kubernetes/blob/8bd5514be14de7fb3105962c8fc206ef5e921a0b/openshift-hack/e2e/annotate/annotate.go).
+3. Help managing the use of disconnected images (i.e. ensure they are only referencing
+   images that are mirrored, at least for conformance tests). See [images used by e2e tests](https://github.com/openshift/origin/blob/cc42164781a728b804d0bb07f09cf878ec8f7807/test/extended/util/image/README.md).
+4. Provide an image annotation or similar mechanism informing where the test
+   binary is embedded in the release payload.
+
+The proposed extensions to `openshift-tests` will require the following changes:
+1. Add extraction logic responsible for retrieving test binaries from the
+   release payload, based on the annotations provided by the authors of these
+   images.
+2. Add logic responsible for iterating over all test binaries invoking the `list`
+   subcommand, and reading the list of tests provided by the binary. Every tests
+   should have an associated binary name, which will be used during invocation.
+3. Expand the current filtering, shuffling, and randomization logic to also take
+   into account the new tests coming from external binaries.
+4. Add logic responsible for iterating over all tests, invoking the `run`
+   subcommand using the associated test's binary and reading its results.
+5. Provide a fallback path allowing to use the vendored tests on demand (`OPENSHIFT_SKIP_EXTERNAL_TESTS`
+   environment variable, for example).
+
+### Risks and Mitigations
+
+- **The current policy when the [origin](https://github.com/openshift/origin/) pull
+  request bringing the updated k8s version breaks the CI is to revert PR. This
+  policy will be harder to execute with this change because reverting the o/k
+  rebase PR can be problematic and was never tried before.**
```

Comment:
What if the kubernetes e2e image rebase will make a current openshift/kubernetes fail? Should we temporary skip the failing tests, do the next k8s rebase and rebase the k8s e2e image?

Reply:
The rebase won't merge until we fix the tests, the goal is to get that signal right away when bringing in newer k8s. The only risk post-merge is that new tests rely on newer kubelet, but that case is already covered in the risks section.